src/main/scala/mesosphere/marathon/core/volume/providers/DVDIProvider.scala

```diff
@@ -0,0 +1,243 @@
+package mesosphere.marathon.core.volume.providers
+
+import com.wix.accord.{ Validator, _ }
+import com.wix.accord.dsl._
+import mesosphere.marathon.core.volume.providers.DVDIProvider._
+import mesosphere.marathon.core.volume.providers.OptionSupport._
+import mesosphere.marathon.state._
+import org.apache.mesos.Protos.Volume.Mode
+import org.apache.mesos.Protos.{ Volume => MesosVolume, CommandInfo, ContainerInfo, Environment }
+
+import scala.collection.JavaConverters._
+
+/**
+  * DVDIProvider (Docker Volume Driver Interface provider) handles external volumes allocated
+  * by a specific docker volume driver plugin. This works for both docker and mesos containerizers,
+  * albeit with some limitations:
+  *   - only a single volume driver per container is allowed when using the docker containerizer
+  *   - docker containerizer requires that referenced volumes be created prior to application launch
+  *   - mesos containerizer only supports volumes mounted in RW mode
+  */
+protected[volume] case object DVDIProvider
+    extends AbstractExternalVolumeProvider("dvdi") with DVDIProviderValidations {
+
+  val driverOption = "dvdi/driver"
+
+  def toMesosVolume(volume: ExternalVolume): MesosVolume =
```

Comment:
make private or even embed it into the method below.

We prefer moving private methods below the public methods that use them.